### ISSUE TITLE
Validate entire schema when get or getAll is called

### ISSUE THREAD

**Issue by samreid** [phetsims/query-string-machine#9]
From #2

**Comment by samreid** [comment]
In https://github.com/phetsims/query-string-machine/issues/2#issuecomment-248013500 @pixelzoom said:

> Assertions sounds most appropriate, since the schema isn't something the user can change. But if it's easier to validate via user-Error checking, go for it.

**Comment by samreid** [comment]
Also the motivation for this was:

> But for testing purposes, defaultValue will only get checked when a value is omitted, so it's possible that there might be a problem with the default value that goes unnoticed. Should the full schema be validated regardless of the parameter string?

**Comment by samreid** [comment]
In the preceding commits, I'm checking that the default value is valid, that the allowedValues are valid and that the default appears in the allowed values (if both provided.)  Error messages are like this:

> Error for query parameter "height": validateSchema.defaultValue: value not allowed: hello, allowedValues: [4,5,6,7,8]

> Error for query parameter "height": validateSchema.allowedValue: should have been a number

@pixelzoom want to take a look?

**Comment by pixelzoom** [comment]
Problems:

(1) None of the error messages identify which key the error pertains to.  E.g. running master of function-builder with `?ea`, I get this rather unhelpful message:

> assert.js:23 Uncaught Error: Assertion failed: default value should be allowed

(2) I cannot find the source of the above error. The 2 schema entries in `FBQueryParameter` that use `defaultValue` look correct.  So perhaps `validateSchema` is buggy?

(3) I only get the above error when assertions are enabled. Isn't validation supposed to occur regardless of whether assertions are enabled?

Back to @samreid.

**Comment by pixelzoom** [comment]
Also wondering about why `window.assert && assert()` is being used, seem atypical and inconsistent.  Should be either `window.assert && window.assert(...)` or `assert && assert(...)`. E.g.

```js
window.assert && assert( ok, 'default value should be allowed' );
```

**Comment by pixelzoom** [comment]
Perhaps the 2 usages of `window.assert && assert` should be `queryStringMachineAssert`?

**Comment by pixelzoom** [comment]
I said:
> (3) I only get the above error when assertions are enabled. Isn't validation supposed to occur regardless of whether assertions are enabled?

No, because this is schema validation, which is not something the user can change. So validating when assertions are enabled is appropriate.

**Comment by pixelzoom** [comment]
I added key name to all error messages in the above commit.  Looks like `validateSchema` must be buggy, because for this entry in FBQueryParameters:

```dev: { type: 'flag' },```

I'm seeing this error message:

```Error for query parameter "dev": defaultValue must be a member of allowedValues```

**Comment by pixelzoom** [comment]
Troubles resolved in above commits.

I still see this TODO item in `validateSchema`, which should be completed before closing this issue:

```// TODO: validate arrays```

Back to @samreid.

**Comment by samreid** [comment]
The code change in validateSchema looks good, I'll move validate arrays to a separate issue.
